### ISSUE TITLE
How layers.Embedding handles out of range indices? 

### ISSUE THREAD

**Issue by grayfall** [keras-team/keras#9264]
I've accidentally created and trained a model with an `Embedding` layer with insufficient input dimension size. The model performs great, but I can't wrap my head around the fact, I've encountered no errors. Here is a small reproducible example: 

```
import numpy as np
from keras.layers import Embedding, GRU, Input, Dense
from keras.models import Model

# create 100 random batches of 32 samples, 128 items each
x = np.random.choice(128, 100*32*128).reshape((100*32, 128))
y = x[:, 0] % 2

samples = Input(batch_shape=(32, 128))
embeddings = Embedding(x.max(), 10)(samples)
rnn = GRU(16)(embeddings)
labels = Dense(1)(rnn)

model = Model(samples, labels)
model.compile(optimizer='Adam', loss='binary_crossentropy')

model.fit(x, y, batch_size=32, epochs=3)
```

When I run this model on a CPU, I get the expected error:

```
InvalidArgumentError (see above for traceback): indices[1,2] = 127 is not in [0, 127)
	 [[Node: embedding_2/Gather = Gather[Tindices=DT_INT32, Tparams=DT_FLOAT, validate_indices=true, _device="/job:localhost/replica:0/task:0/device:CPU:0"](embedding_2/embeddings/read, embedding_2/Cast)]]
```

Yet, this same model trains and performs perfectly on a GPU. As far as I understand the `Embedding` layer calls `backend.gather`, which in its turn calls [`tensorflow::ops::Gather`](https://www.tensorflow.org/api_docs/cc/class/tensorflow/ops/gather) if Keras is using the TF backend (this is the case). This operation's documentation states:

> validate_indices: DEPRECATED. If this operation is assigned to CPU, values in indices are always validated to be within range. If assigned to GPU, out-of-bound indices result in safe but unspecified behavior, which may include raising an error

This explains, why I get the error on a CPU, but I don't quite understand, how it handles the situation on a GPU. I've managed to train a capable model for 30 epochs without ever getting a single error or any performance issues.

I want to make it clear, that I need to understand how it works to use the weights from the model I've trained. It's a fairly heavy model, so I don't want to waste time retraining it from scratch. Basically, I need to understand how the backend maps the out of range indices (there is only one out of range index in my case) to manually rebuild the embedding layer's weights.

**Comment by rex-yue-wu** [comment]
Maybe I am wrong, but my guess is that `gather` simply collects the thing from an indicated memory address without checking whether or not this indicated address within the valid range. This is the reason why you did not see an error message. 

BTW, I think you can reuse your pertained weights directly.

**Comment by grayfall** [comment]
@rex-yue-wu Thanks for the feedback. That's what I believe, too, but it makes me wonder why the model is stable. I've got a frequently occurring out of range index. If the operation indeed grabs values from a memory region that doesn't belong to the tensor, then the corresponding embedding vector is bound to contain garbage and be unstable. I also don't see how it is possible to directly reuse the weights, since there is no way to reconstruct them for that vector.

**Comment by rex-yue-wu** [comment]
The reason why I said you can reuse the weight is because the embedding layer is usually used as the input layer, i.e. the end of chain rule. So it should not affect other weights. 

In the case when out of bound index is received, the model basically updates its corresponding vector in a wrong place, but this wrong place is not in the assigned weights of your embedding layer. So embedding layer is untouched.

What happens if updating a weight in a wrong place? If it is out of your entire model, then why you care about it. Otherwise, it is just add some noise to some weights, and the problematic weights will be pushed to the right direction in the next batch. As long as your model converges, I don't see why you can't use it.

Finally, I think the best way to verify whether or not you can reuse it is to test the model on a small dataset and check its performance.

**Comment by grayfall** [comment]
@rex-yue-wu as I've already mentioned, I can load and use the model on a GPU, but not all of the end-users are going to run inference on a GPU. Consequently, I was going to manually rebuild the model to create a properly sized embedding layer. The layer I have now contains 128 indices (vector weights), while in the end there should be 129, hence I was wondering whether there was any relatively correct way to get the weights for that vector.

**Comment by rex-yue-wu** [comment]
You have many choices. 

1. Only preprocess and postprocess
Given a testing sample X containing 129, simply replace a 129 index with a random index in (0,128) for multiple times, and you have X1, X2, ..., Xk. For each Xk predict its label using your pretrained model, and you have y1, y2, ..., yk. Take the average of all ys and output this prediction.

2. Re-use everything in your pretrained model
You need some custom wrapper layer to explicit handle the extra index, but this is achievable. All you need to train is only the embedding for the new index. For example, you can create two embedding layers inside of this wrapper layer, such that one can directly use weights from pretrained, and the other is the new. But you do need some extra work like `if-else` to control the use of right embedding. 


3. Reuse everything except embedding
Simplest thing you can do is to retrain a model, but only the embedding part, i.e. freeze all other layers except embedding.

**Comment by grayfall** [comment]
@rex-yue-wu I've been thinking somewhere along the lines of the second and third options. As I've mentioned earlier, I'm planning to keep all non-embedding layers intact (reuse their weights) and create a new embedding layer with an extra row of weights, reuse all the other weights and fill-in some values into the new row. Basically, I wanted to get proper weights for that row, but it doesn't seem possible without retraining (or at least significant fine-tuning) the model.

**Comment by rex-yue-wu** [comment]
Since you have one missing embedding, you have to train it anyway. The more weights you freeze, the faster the training is.

**Comment by rex-yue-wu** [comment]
I also believe my first approach (only pre- and post-process) should also work, because 1) only 1 out of 129 index is problematic, 2) you have a sequence of inputs, and you model should already learned how to tolerate the noise added by this problematic index, and 3) using an ensemble model often time improves overall performance.
